x-pack/plugins/observability/server/services/slo/transform_generators/apm_transaction_error_rate.ts

```diff
@@ -0,0 +1,189 @@
+} from '@elastic/elasticsearch/lib/api/types';
+import {
+  SLO_INDEX_TEMPLATE_NAME,
+  SLO_INGEST_PIPELINE_NAME,
+  SLO_RESOURCES_VERSION,
+} from '../../../assets/constants';
+import { getSLOTransformTemplate } from '../../../assets/transform_templates/slo_transform_template';
+import {
+  apmTransactionErrorRateSLOSchema,
+  APMTransactionErrorRateSLO,
+  SLO,
+} from '../../../types/models';
+import { ALL_VALUE } from '../../../types/schema';
+import { TransformGenerator } from '.';
+
+const APM_SOURCE_INDEX = 'metrics-apm*';
+const ALLOWED_STATUS_CODES = ['2xx', '3xx', '4xx', '5xx'];
+const DEFAULT_GOOD_STATUS_CODES = ['2xx', '3xx', '4xx'];
+
+export class ApmTransactionErrorRateTransformGenerator implements TransformGenerator {
+  public getTransformParams(slo: SLO, spaceId: string): TransformPutTransformRequest {
+    if (!apmTransactionErrorRateSLOSchema.is(slo)) {
+      throw new Error(`Cannot handle SLO of indicator type: ${slo.indicator.type}`);
+    }
+
+    return getSLOTransformTemplate(
+      this.buildTransformId(slo),
+      this.buildSource(slo),
+      this.buildDestination(slo, spaceId),
+      this.buildGroupBy(),
+      this.buildAggregations(slo)
+    );
+  }
+
+  private buildTransformId(slo: APMTransactionErrorRateSLO): string {
+    return `slo-${slo.id}`;
+  }
+
+  private buildSource(slo: APMTransactionErrorRateSLO) {
+    const queryFilter = [];
+    if (slo.indicator.params.service !== ALL_VALUE) {
+      queryFilter.push({
+        match: {
+          'service.name': slo.indicator.params.service,
+        },
+      });
+    }
+
+    if (slo.indicator.params.environment !== ALL_VALUE) {
+      queryFilter.push({
+        match: {
+          'service.environment': slo.indicator.params.environment,
+        },
+      });
+    }
+
+    if (slo.indicator.params.transaction_name !== ALL_VALUE) {
+      queryFilter.push({
+        match: {
+          'transaction.name': slo.indicator.params.transaction_name,
+        },
+      });
+    }
+
+    if (slo.indicator.params.transaction_type !== ALL_VALUE) {
+      queryFilter.push({
+        match: {
+          'transaction.type': slo.indicator.params.transaction_type,
+        },
+      });
+    }
+
+    return {
+      index: APM_SOURCE_INDEX,
+      runtime_mappings: {
+        'slo.id': {
+          type: 'keyword' as MappingRuntimeFieldType,
+          script: {
+            source: `emit('${slo.id}')`,
+          },
+        },
+      },
+      query: {
+        bool: {
+          filter: [
+            {
+              match: {
+                'transaction.root': true,
+              },
+            },
+            ...queryFilter,
+          ],
+        },
+      },
+    };
+  }
+
+  private buildDestination(slo: APMTransactionErrorRateSLO, spaceId: string) {
+    if (slo.settings.destination_index === undefined) {
+      return {
+        pipeline: SLO_INGEST_PIPELINE_NAME,
+        index: `${SLO_INDEX_TEMPLATE_NAME}-v${SLO_RESOURCES_VERSION}-${spaceId}`,
```

Comment:
I wonder if we shouldn't have a helper function for this pattern instead of defining it in multiple places? In case we end up having to change the template or add an additional attribute. I'm thinking mostly about someone having to make a change to that pattern and missing one.  Something similar to `getPipelinePrefix()` in the `ResourceInstaller`.

Reply:
Makes sense! I'll change that